consensus/polybft/state.go

```diff
@@ -398,34 +398,82 @@ func (s *State) getExitEventsForProof(epoch, checkpointBlock uint64) ([]*ExitEve
 	})
 }
 
-// getExitEvents returns exit events for given epoch and provided filter
-func (s *State) getExitEvents(epoch uint64, filter func(exitEvent *ExitEvent) bool) ([]*ExitEvent, error) {
-	var events []*ExitEvent
+// updateExitEvents updates the exit events that happened in given block and old epoch to the provided new epoch
+// that means their key in db changes, so we have to remove the old event and reinsert the new one
+func (s *State) updateExitEvents(blockNumber, oldEpoch, newEpoch uint64) error {
+	return s.db.Update(func(tx *bolt.Tx) error {
+		exitEvents, err := getExitEvents(tx, oldEpoch, func(exitEvent *ExitEvent) bool {
```

Comment:
Correct me, if I'm wrong. We are writing events in PostBlock hook, but if it is the last block of the epoch we overwrite event keys in a separate transaction.

Reply:
I think this was also discussed here https://github.com/0xPolygon/polygon-edge/pull/1113#discussion_r1068034498

Reply:
@b00ris In this case `PostBlock` and `PostEpoch` are both called sequentially in the same `OnBlockInserted` call. First the `PostBlock` will insert them in the current epoch, but if it is the end of epoch, `PostEpoch` will update them to new epoch, so we won't be starting a new epoch before doing this (insures that security about finalizing exit events), but yes, we have two `db` transactions, because we didn't want to misuse `PostBlock` to do work of `PostEpoch`. It misses the idea of those two requests that should be separated.

As Ferran mention we discussed this in detail in https://github.com/0xPolygon/polygon-edge/pull/1113#discussion_r1068034498. We don't

Reply:
Ok, so reverted to my first commit, as we discussed on slack: https://github.com/0xPolygon/polygon-edge/pull/1113/commits/a25085b5b42d6b03aeee49817e4dba246299e3ad